README.md
# GitHub Pull Request Monitor

[![Travis](https://img.shields.io/travis/erichbehrens/pull-request-monitor.svg)](https://travis-ci.org/erichbehrens/pull-request-monitor)
[![Marketplace Version](https://vsmarketplacebadge.apphb.com/version/erichbehrens.pull-request-monitor.svg)](https://marketplace.visualstudio.com/items?itemName=erichbehrens.pull-request-monitor)
[![Installs](https://vsmarketplacebadge.apphb.com/installs/erichbehrens.pull-request-monitor.svg)](https://marketplace.visualstudio.com/items?itemName=erichbehrens.pull-request-monitor)
[![Greenkeeper badge](https://badges.greenkeeper.io/erichbehrens/pull-request-monitor.svg)](https://greenkeeper.io/)

This extension uses the GitHub api to monitor the state of your pull requests and let you know when it's time to merge or if someone requested changes.

![Statusbar items](images/statusBarItems.png)

Source code on GitHub: https://github.com/erichbehrens/pull-request-monitor

## Features

- Monitor all the pull requests for your account
- Monitor a specific repository (and quickly switch between them)
- Quickly open GitHub pull requests from VS Code
- Colors and icons to identify pull requests that require attention

### Colors

![color green](images/color-green.png) **Green**: there are no conflicts, build is passing (if any), reviews are approved (if any)

![color red](images/color-red.png) **Red**: opposite of green or pull request closed

![color white](images/color-white.png) **White**: waiting for status

![color violet](images/color-violet.png) **Violet**: merged

### Icons

Note: white icons can become green or red depending on the pull request state.

#### State

![icon](images/icon-state-open.png) Open

![icon](images/icon-state-merged.png) Merged

![icon](images/icon-state-closed.png) Closed

#### Build

![icon](images/icon-build-ok.png) Build passes

![icon](images/icon-build-ko.png) Build fails

#### Branch

![icon](images/icon-mergeable-ok.png) Mergeable

![icon](images/icon-mergeable-ko.png) Conflicts

![icon](images/icon-mergeable-unknown.png) Unknown mergeable state

#### Reviews

![icon](images/icon-reviews-ok.png) Approved reviews

![icon](images/icon-reviews-ko.png) Changes requested

![icon](images/icon-reviews-comments.png) There are comments

## Instructions

- Install the extension

- Generate a GitHub token here: https://github.com/settings/tokens

> **Required permissions:**
>
> If you only need to monitor public repositories enable `public_repo`, if you use private teams enable `repo`.

- Open the command palette and execute `PullRequestMonitor.setToken`

- Paste your token, the extension will start monitoring your pull requests

## Extension Commands

- `PullRequestMonitor.setToken`: set the GitHub token

- `PullRequestMonitor.start`: start monitoring = refresh pull request state every minute

- `PullRequestMonitor.stop`: stop monitoring

- `PullRequestMonitor.refresh`: refresh pull request state

- `PullRequestMonitor.setMode`: select the mode between `viewer` (your pull requests) or `repository`

- `PullRequestMonitor.selectRepository`: select the repository to monitor through the list of your repositories (some private repositories will not appear here, in this case use `PullRequestMonitor.enterRepositoryName` )

- `PullRequestMonitor.enterRepositoryName`: set the private repository name you want to monitor. Something like `your-team-name/awesome-project`

## Extension configuration

- `pullRequestMonitor.refreshInterval`: `number` default = `60`, refresh interval in seconds (min `15`s)

- `pullRequestMonitor.showMerged`: `boolean` default = `false`, show or hide merged pull requests

- `pullRequestMonitor.showClosed` `boolean`, show or hide closed pull requests

- `pullRequestMonitor.autostart` `boolean`, automatically start the extension

- `pullRequestMonitor.count`: `number` default = `6`, number of pull requests to show

- `pullRequestMonitor.githubEnterpriseUrl`: `string` default = `null`, GitHub enterprise url. Leave empty if you don't use GitHub enterprise.

- `pullRequestMonitor.allowUnsafeSSL`: `boolean` default = `false`, allow unsafe certificates for GitHub enterprise

### Default configuration

```json
{
    "pullRequestMonitor.refreshInterval": 60,
    "pullRequestMonitor.showClosed": false,
    "pullRequestMonitor.showMerged": false,
    "pullRequestMonitor.autostart": true,
    "pullRequestMonitor.count": 6,
    "pullRequestMonitor.githubEnterpriseUrl": null,
    "pullRequestMonitor.allowUnsafeSSL": false,
}
```


CHANGELOG.md
# Change Log

## v1.7.1

- Fix configuration setting name

## v1.7.0

- Add support for GitHub enterprise

## v1.6.1

- Security update on vscode dependency

## v1.6.0

- Added setting for number of pull request to show
- Show message when there are no PRs

## v1.5.3

- Bugfix for repositories without tests
- Updated extension icon

## v1.5.2

- Bugfix

## v1.5.1

- Improved network and auth error handling

## v1.4.0

- Autostart extension with VS Code

## v1.3.1

- Update readme with icons

## v1.3.0

- Add configuration for refresh interval and to show/hide closed/merged pull requests
- Major code cleanup

## v1.2.0

- Changed name to *GitHub Pull Request Monitor*
- Use more distinguishable icons for failing tests and conflicts
- Show icon for comments
- Improve showing icons for `approved` and `change requested`

## v1.1.0

- Save user settings
- Rename `setRepository` command to `enterRepositoryName`

## v1.0.4

- Use `isomorphic-fetch` to support systems other than Windows

## v1.0.0

- Initial release


test/index.js
//
// PLEASE DO NOT MODIFY / DELETE UNLESS YOU KNOW WHAT YOU ARE DOING
//
// This file is providing the test runner to use when running extension tests.
// By default the test runner in use is Mocha based.
//
// You can provide your own test runner if you want to override it by exporting
// a function run(testRoot: string, clb: (error:Error) => void) that the extension
// host can call to run the tests. The test runner is expected to use console.log
// to report the results back to the caller. When the tests are finished, return
// a possible error to the callback or null if none.

const testRunner = require('vscode/lib/testrunner');

// You can directly control Mocha options by uncommenting the following lines
// See https://github.com/mochajs/mocha/wiki/Using-mocha-programmatically#set-options for more info
testRunner.configure({
	ui: 'tdd', 		// the TDD UI is being used in extension.test.js (suite, test, etc.)
	useColors: true, // colored output from test results
});

module.exports = testRunner;


test/utils.test.js
const assert = require('assert');
const { getReviewState, getReviewsByAuthor, getLastStateByAuthor } = require('../src/utils');

suite('Utils Tests', () => {
	const TEST_REVIEWS = [
		{ node: { author: { login: 'alice' }, state: 'COMMENTED' } },
		{ node: { author: { login: 'jane' }, state: 'COMMENTED' } },
		{ node: { author: { login: 'john' }, state: 'APPROVED' } },
		{ node: { author: { login: 'jane' }, state: 'APPROVED' } },
		{ node: { author: { login: 'marc' }, state: 'CHANGES_REQUESTED' } },
	];

	test('getReviewsByAuthor', () => {
		const expected = {
			alice: [
				{ author: { login: 'alice' }, state: 'COMMENTED' },
			],
			john: [
				{ author: { login: 'john' }, state: 'APPROVED' },
			],
			jane: [
				{ author: { login: 'jane' }, state: 'COMMENTED' },
				{ author: { login: 'jane' }, state: 'APPROVED' },
			],
			marc: [
				{ author: { login: 'marc' }, state: 'CHANGES_REQUESTED' },
			],
		};
		const actual = getReviewsByAuthor(TEST_REVIEWS);
		assert.deepEqual(actual, expected);
	});

	test('getLastStateByAuthor', () => {
		const expected = ['APPROVED', 'APPROVED', 'CHANGES_REQUESTED'];
		const reviewsByAuthor = getReviewsByAuthor(TEST_REVIEWS);
		const actual = getLastStateByAuthor(reviewsByAuthor);
		assert.deepEqual(actual, expected);
	});

	test('getReviewState - without reviews', () => {
		const reviews = [];
		const expected = {
			hasComments: false,
			isApproved: undefined,
			hasPendingChangeRequests: undefined,
			reviewsPassing: true,
		};
		const actual = getReviewState({ edges: reviews });
		assert.deepEqual(actual, expected);
	});
	test('getReviewState - with changes requested', () => {
		const expected = {
			hasComments: true,
			isApproved: false,
			hasPendingChangeRequests: true,
			reviewsPassing: false,
		};
		const actual = getReviewState({ edges: TEST_REVIEWS });
		assert.deepEqual(actual, expected);
	});

	test('getReviewState - approved', () => {
		const reviews = [
			{ node: { author: { login: 'alice' }, state: 'COMMENTED' } },
			{ node: { author: { login: 'jane' }, state: 'COMMENTED' } },
			{ node: { author: { login: 'john' }, state: 'APPROVED' } },
			{ node: { author: { login: 'jane' }, state: 'APPROVED' } },
			{ node: { author: { login: 'marc' }, state: 'CHANGES_REQUESTED' } },
			{ node: { author: { login: 'marc' }, state: 'APPROVED' } },
		];
		const expected = {
			hasComments: true,
			isApproved: true,
			hasPendingChangeRequests: false,
			reviewsPassing: true,
		};
		const actual = getReviewState({ edges: reviews });
		assert.deepEqual(actual, expected);
	});

	test('getReviewState - without comments', () => {
		const reviews = [
			{ node: { author: { login: 'john' }, state: 'APPROVED' } },
			{ node: { author: { login: 'jane' }, state: 'APPROVED' } },
			{ node: { author: { login: 'marc' }, state: 'CHANGES_REQUESTED' } },
			{ node: { author: { login: 'marc' }, state: 'APPROVED' } },
		];
		const expected = {
			hasComments: false,
			isApproved: true,
			hasPendingChangeRequests: false,
			reviewsPassing: true,
		};
		const actual = getReviewState({ edges: reviews });
		assert.deepEqual(actual, expected);
	});

	test('getReviewState - with comments only', () => {
		const reviews = [
			{ node: { author: { login: 'john' }, state: 'COMMENTED' } },
			{ node: { author: { login: 'jane' }, state: 'COMMENTED' } },
			{ node: { author: { login: 'marc' }, state: 'COMMENTED' } },
			{ node: { author: { login: 'marc' }, state: 'COMMENTED' } },
		];
		const expected = {
			hasComments: true,
			isApproved: undefined,
			hasPendingChangeRequests: undefined,
			reviewsPassing: true,
		};
		const actual = getReviewState({ edges: reviews });
		assert.deepEqual(actual, expected);
	});
});


src/__tests__/utils.js
const { getReviewState, getReviewsByAuthor, getLastStateByAuthor } = require('../utils');

const TEST_REVIEWS = [
	{ node: { author: { login: 'alice' }, state: 'COMMENTED' } },
	{ node: { author: { login: 'jane' }, state: 'COMMENTED' } },
	{ node: { author: { login: 'john' }, state: 'APPROVED' } },
	{ node: { author: { login: 'jane' }, state: 'APPROVED' } },
	{ node: { author: { login: 'marc' }, state: 'CHANGES_REQUESTED' } },
];

test('getReviewsByAuthor', () => {
	const expected = {
		alice: [
			{ author: { login: 'alice' }, state: 'COMMENTED' },
		],
		john: [
			{ author: { login: 'john' }, state: 'APPROVED' },
		],
		jane: [
			{ author: { login: 'jane' }, state: 'COMMENTED' },
			{ author: { login: 'jane' }, state: 'APPROVED' },
		],
		marc: [
			{ author: { login: 'marc' }, state: 'CHANGES_REQUESTED' },
		],
	};
	const actual = getReviewsByAuthor(TEST_REVIEWS);
	expect(actual).toEqual(expected);
});

test('getLastStateByAuthor', () => {
	const expected = ['APPROVED', 'APPROVED', 'CHANGES_REQUESTED'];
	const reviewsByAuthor = getReviewsByAuthor(TEST_REVIEWS);
	const actual = getLastStateByAuthor(reviewsByAuthor);
	expect(actual).toEqual(expected);
});

test('getReviewState - without reviews', () => {
	const reviews = [];
	const expected = {
		hasComments: false,
		isApproved: undefined,
		hasPendingChangeRequests: undefined,
		reviewsPassing: true,
	};
	const actual = getReviewState({ edges: reviews });
	expect(actual).toEqual(expected);
});

test('getReviewState - with changes requested', () => {
	const expected = {
		hasComments: true,
		isApproved: false,
		hasPendingChangeRequests: true,
		reviewsPassing: false,
	};
	const actual = getReviewState({ edges: TEST_REVIEWS });
	expect(actual).toEqual(expected);
});

test('getReviewState - approved', () => {
	const reviews = [
		{ node: { author: { login: 'alice' }, state: 'COMMENTED' } },
		{ node: { author: { login: 'jane' }, state: 'COMMENTED' } },
		{ node: { author: { login: 'john' }, state: 'APPROVED' } },
		{ node: { author: { login: 'jane' }, state: 'APPROVED' } },
		{ node: { author: { login: 'marc' }, state: 'CHANGES_REQUESTED' } },
		{ node: { author: { login: 'marc' }, state: 'APPROVED' } },
	];
	const expected = {
		hasComments: true,
		isApproved: true,
		hasPendingChangeRequests: false,
		reviewsPassing: true,
	};
	const actual = getReviewState({ edges: reviews });
	expect(actual).toEqual(expected);
});

test('getReviewState - without comments', () => {
	const reviews = [
		{ node: { author: { login: 'john' }, state: 'APPROVED' } },
		{ node: { author: { login: 'jane' }, state: 'APPROVED' } },
		{ node: { author: { login: 'marc' }, state: 'CHANGES_REQUESTED' } },
		{ node: { author: { login: 'marc' }, state: 'APPROVED' } },
	];
	const expected = {
		hasComments: false,
		isApproved: true,
		hasPendingChangeRequests: false,
		reviewsPassing: true,
	};
	const actual = getReviewState({ edges: reviews });
	expect(actual).toEqual(expected);
});

test('getReviewState - with comments only', () => {
	const reviews = [
		{ node: { author: { login: 'john' }, state: 'COMMENTED' } },
		{ node: { author: { login: 'jane' }, state: 'COMMENTED' } },
		{ node: { author: { login: 'marc' }, state: 'COMMENTED' } },
		{ node: { author: { login: 'marc' }, state: 'COMMENTED' } },
	];
	const expected = {
		hasComments: true,
		isApproved: undefined,
		hasPendingChangeRequests: undefined,
		reviewsPassing: true,
	};
	const actual = getReviewState({ edges: reviews });
	expect(actual).toEqual(expected);
});


src/utils.js
exports.getCommitIcon = (value) => {
	// https://developer.github.com/v4/reference/enum/statusstate/
	if (!value) return undefined;
	switch (value.state) {
		case 'SUCCESS': return '$(check)';
		case 'PENDING': return '$(kebab-horizontal)';
		case 'FAILURE': return '$(tools)';
	}
};

exports.getMergeableIcon = (value) => {
	// https://developer.github.com/v4/reference/enum/statusstate/
	switch (value) {
		case 'MERGEABLE': return '$(git-merge)';
		case 'UNKNOWN': return '$(question)';
		case 'CONFLICTING': return '$(alert)';
	}
};

exports.getPullRequestStateIcon = (value) => {
	// https://developer.github.com/v4/reference/enum/pullrequeststate/
	switch (value) {
		case 'MERGED': return '$(git-merge)';
		case 'OPEN': return '$(git-pull-request)';
		case 'CLOSED': return '$(x)';
	}
};

exports.getColor = (mergeableState) => {
	switch (mergeableState) {
		case 'MERGED': return 'rgba(144, 0, 192, 1)'; // #9000c0
		case 'MERGEABLE': return 'rgba(64, 255, 16, 1)'; // #40ff10
		case 'CLOSED':
		case 'FAILURE': return 'rgba(144, 0, 0, 1)'; // #900000
		default: return 'rgba(255, 255, 255, 1)';
	}
};

exports.getMergeableState = (pr, reviews) => {
	// https://developer.github.com/v4/reference/enum/mergeablestate/
	const commit = pr.commits.nodes[0].commit.status;
	const { potentialMergeCommit } = pr;
	if (['MERGED', 'CLOSED'].includes(pr.state)) {
		return pr.state;
	}
	if (
		pr.mergeable === 'MERGEABLE' &&
		reviews &&
		(commit === null /* no tests defined */ || commit.state === 'SUCCESS') &&
		potentialMergeCommit && potentialMergeCommit.status === null
	) {
		return 'MERGEABLE';
	}
	if (
		pr.mergeable === 'CONFLICTING' ||
		!reviews ||
		(commit !== null /* tests definied */ && commit.state === 'FAILURE')
	) {
		return 'FAILURE';
	}
	return 'OPEN';
};

exports.getStatesFilter = (showMerged, showClosed) => {
	if (showMerged && showClosed) {
		return ''; // no filter apllied
	}
	const states = ['OPEN'];
	if (showClosed) {
		states.push('CLOSED');
	}
	if (showMerged) {
		states.push('MERGED');
	}
	return `states: [${states.join(' ')}]`;
};

function getReviewsByAuthor(reviews) {
	return reviews.reduce((acc, { node }) => {
		acc[node.author.login] = acc[node.author.login] || [];
		acc[node.author.login].push(node);
		return acc;
	}, {});
}

function getLastStateByAuthor(reviewsByAuthor) {
	return Object.keys(reviewsByAuthor).map((author) => {
		let lastState;
		reviewsByAuthor[author].forEach(({ state }) => {
			if (['CHANGES_REQUESTED', 'APPROVED'].includes(state)) lastState = state;
		});
		return lastState;
	}).filter(item => item);
}

exports.getReviewState = (reviews) => {
	const reviewsCount = reviews.edges.length;
	let hasPendingChangeRequests;
	let isApproved;
	if (reviewsCount > 0) {
		const reviewsByAuthor = getReviewsByAuthor(reviews.edges);
		const lastStateByAuthor = getLastStateByAuthor(reviewsByAuthor);
		if (lastStateByAuthor && lastStateByAuthor.length > 0) {
			hasPendingChangeRequests = lastStateByAuthor.some(state => state === 'CHANGES_REQUESTED');
			isApproved = lastStateByAuthor.every(state => state === 'APPROVED');
		}
	}
	const hasComments = reviews.edges.some(({ node }) => node.state === 'COMMENTED');
	const reviewsPassing = reviewsCount === 0 || !hasPendingChangeRequests || isApproved;
	return {
		reviewsPassing,
		hasComments,
		hasPendingChangeRequests,
		isApproved,
	};
};

exports.getEndpointUrl = (githubEnterpriseUrl) => {
	if (githubEnterpriseUrl) {
		return `${githubEnterpriseUrl}/api/graphql`;
	}
	return 'https://api.github.com/graphql';
};

/* test exports */
exports.getReviewsByAuthor = getReviewsByAuthor;
exports.getLastStateByAuthor = getLastStateByAuthor;


src/queries.js
exports.viewer = `query {
  viewer {
    login
      pullRequests(last: @count @states) {
          nodes {
            repository{
              name
            }
            number
            mergeable
            state
            title
            mergedAt
            merged
            url
            potentialMergeCommit {
              status {
                state
                commit {
                  status{ state }
                }
              }
            }
            commits(last: 1){
              nodes{
                commit{
                  status{
                    state
                  }
                }
              }
            }
            reviews(first: 10) {
              edges {
                node {
                  state
                  author {
                    login
                  }
                }
              }
            }
          }
    }
  }
}`;

exports.repository = `{
  repository(owner: "@owner" name: "@name") {
    pullRequests(last: @count @states) {
      nodes {
        repository{
          name
        }
        number
        mergeable
        state
        title
        mergedAt
        merged
        url
        potentialMergeCommit {
          status {
            state
            commit {
              status{ state }
            }
          }
        }
        commits(last: 1){
          nodes{
            commit{
              status{
                state
              }
            }
          }
        }
        reviews(first: 10) {
          edges {
            node {
              state
              author {
                login
              }
            }
          }
        }
      }
    }
  }
}`;

exports.repositories = `query {
  viewer {
    login
    repositories(first: 100) {
      nodes{
        name
        nameWithOwner
        owner {
          login
        }
      }
    }
  }
}`;


src/requests.js
const { window } = require('vscode'); // eslint-disable-line import/no-unresolved
const fetch = require('isomorphic-fetch');
const https = require('https');
const queries = require('./queries');
const { getStatesFilter } = require('./utils');

let errorCount = 0;

async function execQuery(token, query, showError, graphqlEndpoint, allowUnsafeSSL = false) {
	if (!graphqlEndpoint.startsWith('https://') && showError) {
		window.showErrorMessage('GitHub enterprise url must start with https://');
		return { status: 'error' };
	}
	if (showError) {
		errorCount = 0;
	}
	if (!token) {
		window.showWarningMessage('Pull Request Monitor needs a token');
		return { status: 'error' };
	}
	try {
		const res = await fetch(graphqlEndpoint, {
			method: 'POST',
			headers: { Authorization: `bearer ${token}` },
			body: JSON.stringify({ query }),
			agent: new https.Agent({ rejectUnauthorized: !allowUnsafeSSL }),
		});
		if (res.status === 200) {
			const { data } = await res.json();
			return { status: 'ok', data };
		}
		if (res.status === 401 || res.status === 403) {
			window.showErrorMessage('Pull Request Monitor token not authorized');
			return { status: 'error', code: 401 };
		}
		window.showErrorMessage(`Pull Request Monitor http error ${res.status}`);
		return { status: 'error', code: res.status };
	} catch (e) {
		console.error(e); // eslint-disable-line no-console
		if (!showError) {
			errorCount += 1;
		}
		if (showError || errorCount === 2) {
			window.showErrorMessage('Pull Request Monitor error fetching data');
		}
		return { status: 'error' };
	}
}

exports.loadPullRequests =
	async (token,
		{ mode, showMerged, showClosed, repository, showError, count, url, allowUnsafeSSL },
	) => {
		let query = queries[mode]
			.replace('@states', getStatesFilter(showMerged, showClosed))
			.replace('@count', count);
		if (mode === 'repository') {
			if (!repository) {
				window.showWarningMessage('Pull Request Monitor needs a repository to watch');
				return { status: 'error' };
			}
			query = query.replace('@owner', repository.owner).replace('@name', repository.name);
		}
		const { status, code, data } = await execQuery(token, query, showError, url, allowUnsafeSSL);
		const pullRequests = data && data[mode].pullRequests.nodes;
		return { status, code, data: pullRequests };
	};

exports.loadRepositories = async (token, { url, allowUnsafeSSL }) => {
	const query = queries.repositories;
	const { status, code, data } = await execQuery(token, query, true, url, allowUnsafeSSL);
	const repositories = data && data.viewer.repositories.nodes;
	return { status, code, data: repositories };
};


src/extension.js
const vscode = require('vscode'); // eslint-disable-line import/no-unresolved
const { clearTimeout, setTimeout } = require('timers');
const { getCommitIcon, getColor, getMergeableIcon, getMergeableState, getPullRequestStateIcon, getReviewState, getEndpointUrl } = require('./utils');
const { loadPullRequests, loadRepositories } = require('./requests');

const MODES = {
	VIEWER: 'viewer',
	REPOSITORY: 'repository',
};

let statusBarItems;

let timer;

function createStatusBarItem(context, prId, url) {
	const statusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
	context.subscriptions.push(statusBarItem);
	const disposable = vscode.commands.registerCommand(`PullRequestMonitor.openPullRequest.${prId}`, () => {
		vscode.commands.executeCommand('vscode.open', vscode.Uri.parse(url));
	});
	context.subscriptions.push(disposable);
	statusBarItems[prId] = statusBarItem;
}

let pullRequests = [];
let refreshButton;
let noResultsLabel;

async function getPullRequests(context, showError) {
	clearTimeout(timer);
	let overrideRefreshInterval;
	try {
		const mode = context.globalState.get('mode', MODES.VIEWER);
		const repository = context.globalState.get('currentRepository');
		const showMerged = vscode.workspace.getConfiguration('pullRequestMonitor').get('showMerged');
		const showClosed = vscode.workspace.getConfiguration('pullRequestMonitor').get('showClosed');
		const count = vscode.workspace.getConfiguration('pullRequestMonitor').get('count');
		// configure the URL settings
		const url = getEndpointUrl(vscode.workspace.getConfiguration('pullRequestMonitor').get('githubEnterpriseUrl'));
		// configure SSL
		const allowUnsafeSSL = vscode.workspace.getConfiguration('pullRequestMonitor').get('allowUnsafeSSL');
		const updatedPullRequests = await loadPullRequests(context.globalState.get('token'), { mode, showMerged, showClosed, repository, showError, count, url, allowUnsafeSSL });
		if (updatedPullRequests.code === 401) {
			refreshButton.command = 'PullRequestMonitor.setToken';
			refreshButton.text = '$(key)';
			refreshButton.tooltip = 'Set GitHub token for Pull Request Monitor';
			return;
		}
		if (updatedPullRequests.status === 'error') {
			refreshButton.command = 'PullRequestMonitor.refresh.showError';
			refreshButton.text = '$(zap)';
			refreshButton.tooltip = 'Connect Pull Request Monitor';
			overrideRefreshInterval = 15;
		} else {
			refreshButton.command = 'PullRequestMonitor.refresh';
			refreshButton.text = '$(sync)';
			refreshButton.tooltip = 'Refresh Pull Request Monitor';
			pullRequests = updatedPullRequests.data;
		}
		if (!statusBarItems) {
			statusBarItems = {};
		} else {
			const prIds = pullRequests.map(pr => `${pr.repository.name}${pr.number}`);
			Object.keys(statusBarItems)
				.forEach(item => !prIds.includes(item) && statusBarItems[item].hide());
		}
		if (pullRequests.length === 0) {
			noResultsLabel.show();
		} else {
			noResultsLabel.hide();
		}
		pullRequests.forEach((pr) => {
			const prId = `${pr.repository.name}${pr.number}`;
			if (!statusBarItems[prId]) {
				createStatusBarItem(context, prId, pr.url);
			}
			const {
				reviewsPassing,
				hasComments,
				hasPendingChangeRequests,
				isApproved,
			} = getReviewState(pr.reviews);
			const mergeableState = getMergeableState(pr, reviewsPassing);
			const closed = mergeableState === 'CLOSED';

			const statusBarItem = statusBarItems[prId];
			const text = [
				getPullRequestStateIcon(pr.state),
				pr.number,
				!pr.merged && !closed && getCommitIcon(pr.commits.nodes[0].commit.status),
				!pr.merged && !closed && getMergeableIcon(pr.mergeable),
				hasComments && '$(comment)',
				hasPendingChangeRequests && '$(thumbsdown)',
				isApproved && '$(thumbsup)',
			];
			statusBarItem.text = text.filter(item => item).join(' ');
			statusBarItem.color = getColor(mergeableState);
			statusBarItem.command = `PullRequestMonitor.openPullRequest.${prId}`;
			statusBarItem.tooltip = pr.title;
			statusBarItem.prominentBackground = true;
			statusBarItem.show();
		});
	} catch (e) {
		console.error(e); // eslint-disable-line no-console
		vscode.window.showErrorMessage('Pull Request Monitor error rendering');
	}
	// we store the interval in seconds and prevent the users to set a value lower than 15s
	const userRefreshInterval = overrideRefreshInterval || Number.parseInt(vscode.workspace.getConfiguration('pullRequestMonitor').get('refreshInterval'), 10);
	const refreshInterval = userRefreshInterval >= 15 ? userRefreshInterval : 60;
	if (!refreshInterval) return;
	timer = setTimeout(() => getPullRequests(context), refreshInterval * 1000);
}

function setRepository(context, nameWithOwner) {
	const currentRepository = context.globalState.get('currentRepository');
	if (nameWithOwner && (!currentRepository || currentRepository.nameWithOwner !== nameWithOwner)) {
		context.globalState.update('currentRepository', {
			nameWithOwner,
			owner: nameWithOwner.split('/')[0],
			name: nameWithOwner.split('/')[1],
		});
		context.globalState.update('mode', MODES.REPOSITORY);
		getPullRequests(context);
	}
}

function activate(context) {
	noResultsLabel = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
	noResultsLabel.command = 'PullRequestMonitor.selectRepository';
	noResultsLabel.text = 'No PRs';
	noResultsLabel.tooltip = 'Select another repository';
	context.subscriptions.push(noResultsLabel);

	refreshButton = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
	refreshButton.command = 'PullRequestMonitor.refresh';
	refreshButton.text = '$(sync)';
	refreshButton.tooltip = 'Refresh Pull Request Monitor';
	refreshButton.show();
	context.subscriptions.push(refreshButton);

	let disposable = vscode.commands.registerCommand('PullRequestMonitor.start', (options = {}) => {
		const { silent } = options;
		getPullRequests(context);
		if (!silent) vscode.window.showInformationMessage('Pull Request Monitor started!');
	});

	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('PullRequestMonitor.stop', () => {
		clearTimeout(timer);
		vscode.window.showInformationMessage('Pull Request Monitor stopped!');
	});

	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('PullRequestMonitor.refresh', () => {
		getPullRequests(context);
		vscode.window.showInformationMessage('Pull Request Monitor refreshing');
	});

	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('PullRequestMonitor.refresh.showError', () => {
		getPullRequests(context, true);
		vscode.window.showInformationMessage('Pull Request Monitor refreshing');
	});

	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('PullRequestMonitor.setMode', async () => {
		const selectedMode = await vscode.window.showQuickPick([MODES.REPOSITORY, MODES.VIEWER], { placeHolder: 'Please select the mode' });
		if (selectedMode && context.globalState.get('mode') !== selectedMode) {
			if (selectedMode === MODES.REPOSITORY && !context.globalState.get('currentRepository')) {
				vscode.commands.executeCommand('PullRequestMonitor.selectRepository');
				return;
			}
			context.globalState.update('mode', selectedMode);
			getPullRequests(context);
		}
	});

	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('PullRequestMonitor.selectRepository', async () => {
		const url = getEndpointUrl(vscode.workspace.getConfiguration('pullRequestMonitor').get('githubEnterpriseUrl'));
		const allowUnsafeSSL = vscode.workspace.getConfiguration('pullRequestMonitor').get('allowUnsafeSSL');
		const { data: repositories } = await loadRepositories(context.globalState.get('token'), { url, allowUnsafeSSL });
		if (!repositories) {
			return;
		}
		const repositoryNames = repositories.map(repository => repository.nameWithOwner);
		const selectedRepository = await vscode.window.showQuickPick(repositoryNames, { placeHolder: 'Please select the repository' });
		setRepository(context, selectedRepository);
	});

	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('PullRequestMonitor.enterRepositoryName', async () => {
		const selectedRepository = await vscode.window.showInputBox({ placeHolder: 'Please enter the full repository name, ie: your-team/awesome-project' });
		setRepository(context, selectedRepository);
	});

	context.subscriptions.push(disposable);

	disposable = vscode.commands.registerCommand('PullRequestMonitor.setToken', async () => {
		const token = await vscode.window.showInputBox({ placeHolder: 'Please enter your GitHGub token' });
		if (token) {
			context.globalState.update('token', token);
			getPullRequests(context);
			vscode.window.showInformationMessage('Token saved.');
		}
	});

	context.subscriptions.push(disposable);

	if (vscode.workspace.getConfiguration('pullRequestMonitor').get('autostart') && context.globalState.get('token')) {
		vscode.commands.executeCommand('PullRequestMonitor.start', { silent: true });
	}
}
exports.activate = activate;

function deactivate() {
	// todo
}
exports.deactivate = deactivate;
